### DISCUSSION TITLE
How to update a .cime machine port to v3 of config_machines?

### DISCUSSION THREAD

**Discussion by billsacks** [ESMCI/cime#4587] [Q&A] [answered]
When trying to run a case with a recent version of CIME on my Mac, where I have a machine port in my `.cime` directory, I get this error:

```
ERROR: Command: '/usr/bin/xmllint --xinclude --noout --schema /Users/sacks/cesm_code/cesm/cime/CIME/data/config/xml_schemas/config_machines_version3.xsd /Users/sacks/.cime/config_machines.xml' failed with error '/Users/sacks/.cime/config_machines.xml:12: element NODENAME_REGEX: Schemas validity error : Element 'NODENAME_REGEX': This element is not expected. Expected is one of ( NODE_FAIL_REGEX, MPIRUN_RETRY_REGEX, MPIRUN_RETRY_COUNT, OS ).
/Users/sacks/.cime/config_machines.xml fails to validate' from dir '/Users/sacks/cesm_code/cesm/cime/scripts'
```

Poking around, I see that there has been a rework in where NODENAME_REGEX is specified. I can get this to work by removing NODENAME_REGEX from my config_machines.xml file in .cime and adding these diffs in CESM's ccs_config:

```diff
diff --git a/machines/config_machines.xml b/machines/config_machines.xml
index 2c0a847..0f8846d 100644
--- a/machines/config_machines.xml
+++ b/machines/config_machines.xml
@@ -58,6 +58,7 @@ This allows using a different mpirun command to launch unit tests
     <value MACH="coeus">(login[1,2].cluster|compute[0-9]*.cluster)</value>
     <value MACH="derecho">$ENV{NCAR_HOST}:derecho</value>
     <value MACH="frontera">.*frontera</value>
+    <value MACH="green">cgdm-green</value>
     <value MACH="greenplanet-sky24">gplogin\d.gp.local</value>
     <value MACH="gust">gu.*.hpc.ucar.edu</value>
     <value MACH="izumi">^i.*\.ucar\.edu</value>
```

But is there a way to do this without modifying the central version of config_machines – by adding NODENAME_REGEX in some new place in my .cime directory?

Also note that, when I tried this without the above diff – even if I explicitly specified `--machine green` in my create_newcase command – I got this error:

```
Traceback (most recent call last):
  File "/Users/sacks/cesm_code/cesm/cime/scripts/./create_newcase", line 18, in <module>
    _main_func()
  File "/Users/sacks/cesm_code/cesm/cime/CIME/scripts/create_newcase.py", line 440, in _main_func
    case.create(
  File "/Users/sacks/cesm_code/cesm/cime/CIME/case/case.py", line 2479, in create
    self.configure(
  File "/Users/sacks/cesm_code/cesm/cime/CIME/case/case.py", line 1350, in configure
    probed_machine = machobj.probe_machine_name()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sacks/cesm_code/cesm/cime/CIME/XML/machines.py", line 183, in probe_machine_name
    machine = self._probe_machine_name_one_guess(nametomatch)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sacks/cesm_code/cesm/cime/CIME/XML/machines.py", line 213, in _probe_machine_name_one_guess
    return self._probe_machine_name_one_guess_v3(nametomatch)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sacks/cesm_code/cesm/cime/CIME/XML/machines.py", line 306, in _probe_machine_name_one_guess_v3
    return machine
           ^^^^^^^
UnboundLocalError: cannot access local variable 'machine' where it is not associated with a value
```

**Comment by jedwards4b** [comment]
The first part is a feature NODENAME_REGEX should no longer be in your .cime/config_machines.xml.
The second looks like a bug, using the --machine should work - can you open an issue on that.

**Reply by billsacks** [reply]
Thanks @jedwards4b . I just opened #4588 .

Is there a new way for users maintaining a machine port like this to specify an auto-match, or will the `--machine` argument now be required in this case?

**Reply by jedwards4b** [reply]
I think that there isn't currently a new way but we can come up with something.

**Reply by billsacks** [reply]
Thanks. For me personally this isn't a big deal – I can explicitly specify the machine – so we could wait until others ask for this.

That said, one idea that comes to mind is based on the ability to specify defaults for some variables other than just machine configuration variables in your .cime directory. (I forget how this works because I don't use it myself, but I remember it being used for things like the default job email options.) Could this be used to allow a user to specify a default `--machine` argument for create_newcase?

**Reply by billsacks** [reply]
I also noticed that @jasonb5 did something with the [docker port](https://github.com/ESMCI/cime/tree/master/docker/.cime) that seemed related to this, but I couldn't figure out how to apply the changes like this to my own .cime directory in a way that worked.

**Reply by jasonb5** [reply]
The change I made basically keeps both v2 and v3 of config_machines in `~/.cime`. The containers [entrypoint](https://github.com/ESMCI/cime/blob/46814b892c92200e77e5977f5b99eb160da8ffd1/docker/entrypoint.sh#L107-L108) will enable the correct one depending on the model. This was needed for E3SM specific system tests as the code base is using v2 while CESM is using v3.

**Reply by billsacks** [reply]
Thanks, @jasonb5 ! This helped me see how it's still possible to have auto-detection of a machine. The latest version in https://github.com/billsacks/mac_cime_configuration (billsacks/mac_cime_configuration@3eafe1147282dbaaa32c68ac4641d1824e84cc2a) allows this auto-detection, following what you've done for the docker machine.

**Reply by billsacks** [reply, marked as answer]
And here are the net changes I needed to make:

https://github.com/billsacks/mac_cime_configuration/compare/64bda4a..3eafe11

It's hard to tell from the diffs, but this consisted of:

(1) Removing the NODENAME_REGEX element from config_machines.xml.

(2) Moving config_machines.xml to MACHINE_NAME/config_machines.xml (in my case, green/config_machines.xml)

(3) Introducing a new top-level config_machines.xml with just the NODENAME_REGEX:

```xml
<?xml version="1.0"?>

<config_machines version="3.0">
	<NODENAME_REGEX>
		<value MACH="green">cgdm-green</value>
	</NODENAME_REGEX>
</config_machines>
```

**Reply by jasonb5** [reply]
@billsacks I'll be doing an overhaul on the documentation soon and will make sure this is included.

**Reply by billsacks** [reply]
Awesome, thanks @jasonb5 !

**Reply by jedwards4b** [reply]
Thanks for figuring it out.

**Reply by billsacks** [reply]
Hmmm, for some reason this solution with a top-level config_machines.xml (just specifying NODENAME_REGEX) and the "real" config_machines entry under a "green" subdirectory is no longer working for me: I'm getting:

```
ERROR: Command: '/usr/bin/xmllint --xinclude --noout --schema /Users/sacks/cism-code/cism-wrapper/cism-wrapper1/cime/CIME/data/config/xml_schemas/config_machines.xsd /Users/sacks/.cime/config_machines.xml' failed with error '/Users/sacks/.cime/config_machines.xml:4: element NODENAME_REGEX: Schemas validity error : Element 'NODENAME_REGEX': This element is not expected. Expected is ( machine ).
/Users/sacks/.cime/config_machines.xml fails to validate' from dir '/Users/sacks/cism-code/cism-wrapper/cism-wrapper1/cime/scripts'
```

And if I remove the top-level config_machines.xml and specify the machine explicitly, I get:

```
ERROR: No machine green found
```

I'm not sure why this was working for me before but is no longer working... I'm probably using slightly different CIME versions than before, but I've tried a few versions this morning. So for now I'm backing out the ability to auto-detect the machine and just putting the "real" config_machines.xml entry at the top level of my `.cime`... so I'm good for now, but @jasonb5 I'd be happy to hear if you have any insights.

**Reply by jedwards4b** [reply]
I just tried your instructions on my laptop @billsacks  and they work fine.  I tried with cime master and with the  cime6.0.217_httpsbranch02 tag.

**Reply by billsacks** [reply]
Well, now it works for me, too. No idea why it stopped working for me for some time, but I'll chalk it up to user error on my part. Sorry for the confusion.

**Reply by billsacks** [reply]
So https://github.com/billsacks/mac_cime_configuration/tree/330157b16ec4ae1476a6f4b361d81c1e8ea57f40 shows what is working for me.

**Comment by adrifoster** [comment]
Hi I'm chiming in because I am now having trouble running things on my personal computer. Before this worked fine, but now I get this error when trying to run the fortran unit tests. I haven't tried running a regular model run yet, but perhaps that will also not work?

```
ERROR: Could not initialize machine object from /Users/afoster/Documents/ncar/CTSM/ccs_config/machines/config_machines.xml, /Users/afoster/.cime/config_machines.xml. This machine is not available for the target CIME_MODEL.
```

I think this is now expected behavior? I know there are workarounds for this (e.g. setting a `CIME_MACHINE` environment variable), but this is less than ideal in terms of portability. 

Is this something that can be worked on?

**Reply by jedwards4b** [reply]
Did you try the changes recommended by Bill Sacks above?

**Comment by adrifoster** [comment]
No, I thought those didn't work any more. I will try those.

**Reply by adrifoster** [reply]
Okay that worked! Thanks for helping! @billsacks I guess just move things back to what you did before?

**Comment by jedwards4b** [comment]
My proposal would be to look for a hierarchy of dot files - .cesm3, .cesm2, .cime for cesm and .e3sm3, .e3sm2, .cime for e3sm - (ufs would follow the same kind of pattern) - thoughts?
